### PR TITLE
Add frontend download capabilities to FrontendScreen and rework permissions

### PR DESCRIPTION
**Pull request by TimoPtr**
## Summary

Introduce `FrontendDownloadManager` to handle file downloads from the WebView, supporting blob URLs, HTTP/HTTPS URLs (via the system `DownloadManager`), and data URIs. Rework the JS bridge, permission manager, and message handler to support the new download flow, and make `DataUriDownloadManager` injectable. Rename `FrontendNavigationEvent` to `FrontendEvent` to be more generic than navigating to handle one time thing happening.

The rework of the whole permission system allow for a better abstraction from the VM point of view and make a generic handling in the screen. It means that it is going to require less work in future PR to request a permission mostly focused only within the PermissionManager and the code that actually need the permission instead of the screen.

## Checklist

- [x] New or updated tests have been added to cover the changes following the testing [guidelines](https://developers.home-assistant.io/docs/android/testing/introduction).
- [x] The code follows the project's [code style](https://developers.home-assistant.io/docs/android/codestyle) and [best_practices](https://developers.home-assistant.io/docs/android/best_practices).
- [x] The changes have been thoroughly tested, and edge cases have been considered.
- [x] Changes are backward compatible whenever feasible. Any breaking changes are documented in the changelog for users and/or in the code for developers depending on the relevance.

## Any other notes

I started working on this quite some time ago and I had to go over multiple iterations before being able to open the PR. It means that maybe there are some leftover from previous iterations.

My plan later is to properly draw in the documentation how things are now communicating and explaining the goal of each stream, like events VS state.

### CODE REVIEW

File: app/src/main/kotlin/io/homeassistant/companion/android/frontend/handler/FrontendMessageHandler.kt [hunk 1]
```diff
@@ -55,11 +57,12 @@ class FrontendMessageHandler @Inject constructor(
     private val threadManager: ThreadManager,
     private val appVersionProvider: AppVersionProvider,
     private val sessionManager: ServerSessionManager,
+    private val downloadManager: FrontendDownloadManager,
     @param:IsAutomotive private val isAutomotive: Boolean,
 ) : FrontendJsHandler,
     FrontendBusObserver {
 
-    private val authResultsFlow = MutableSharedFlow<FrontendHandlerEvent>(extraBufferCapacity = 1)
+    private val jsCallbackEvents = MutableSharedFlow<FrontendHandlerEvent>(extraBufferCapacity = 1)
```
Comment (TimoPtr) [inline app/src/main/kotlin/io/homeassistant/companion/android/frontend/handler/FrontendMessageHandler.kt:65]: I renamed it in the first iteration I've made where I was using this from the handleBlob callback that doesn't exist anymore. I didn't revert because I think it might be better for the usability within the class. Happy to discuss if we want to revert that.

File: app/src/main/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManager.kt [hunk 2]
```diff
@@ -140,37 +168,82 @@ internal class PermissionManager @Inject constructor(
...
+     * Must only be called when a request is in-flight; triggers [FailFast] otherwise.
+     */
+    fun clearPendingPermissionRequest() {
+        FailFast.failWhen(_pendingPermissionRequest.value == null) {
+            "Should have in-flight permission while calling clearPendingPermissionRequest"
         }
+        _pendingPermissionRequest.clear()
+    }
+
+    /**
```
Comment (TimoPtr) [inline app/src/main/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManager.kt:219]: I moved the function in the file to have a better organization I didn't change this method and the next one.

File: app/src/main/kotlin/io/homeassistant/companion/android/frontend/download/FrontendDownloadManager.kt [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+                    }
+                }
+
+                request.addRequestHeader("Cookie", CookieManager.getInstance().getCookie(url))
+            } catch (e: CancellationException) {
+                throw e
+            } catch (e: Exception) {
+                Timber.e(e, "Unable to prepare request")
+            }
+
```
[the remark this answers is not in the document]
Comment (TimoPtr) [reply app/src/main/kotlin/io/homeassistant/companion/android/frontend/download/FrontendDownloadManager.kt]: @jpelgrom the old implementation is not checking for nullable, is it something you ever expected?

File: app/src/main/kotlin/io/homeassistant/companion/android/webview/WebViewContentScreen.kt [hunk 4]
```diff
@@ -126,7 +126,10 @@ internal fun WebViewContentScreen(
         }
         if (webViewInitialized && shouldAskNotificationPermission && supportsNotificationPermission) {
             @SuppressLint("InlinedApi")
-            NotificationPermissionPrompt(onNotificationPermissionResult)
+            NotificationPermissionPrompt(
+                onPermissionResult = onNotificationPermissionResult,
+                onDismiss = {},
```
[the remark this answers is not in the document]
Comment (TimoPtr) [reply app/src/main/kotlin/io/homeassistant/companion/android/webview/WebViewContentScreen.kt:131]: It is the expected behavior discard means that next time we open the app we are going to show again the bottom sheet.

File: app/src/main/kotlin/io/homeassistant/companion/android/frontend/download/FrontendDownloadManager.kt [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+                        const blob = await response.blob();
+                        const reader = new FileReader();
+                        reader.onloadend = $blobCallback;
+                        reader.readAsDataURL(blob);
+                    } catch (e) {
+                        console.error('Blob download failed:', e);
+                    }
+                })();
+        """.trimIndent()
+        externalBusRepository.evaluateScript(jsCode)
```
Comment (jpelgrom) [inline app/src/main/kotlin/io/homeassistant/companion/android/frontend/download/FrontendDownloadManager.kt:177]: This feels like a hack as evaluating raw Javascript functions has little to do with the external bus. This should be discouraged.

File: app/src/main/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManager.kt [hunk 6]
```diff
@@ -140,37 +176,82 @@ internal class PermissionManager @Inject constructor(
...
-
-        val allGrantedResources = pending.alreadyGrantedResources + newlyGrantedResources
-
-        if (allGrantedResources.isNotEmpty()) {
-            Timber.d("Granting WebView resources: $allGrantedResources")
-            pending.webViewRequest.grant(allGrantedResources.toTypedArray())
-        } else {
-            Timber.d("User denied all requested permissions, denying WebView request")
-            pending.webViewRequest.deny()
+    suspend fun requiresStoragePermissionForDownload(onGranted: () -> Unit): Boolean {
```
Comment (jpelgrom) [inline app/src/main/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManager.kt]: Comparing this to `checkNotificationPermission`, I see quite some similarities but also differences (e.g. you can pass a callback here). The names of the functions do not make clear at a glance what the difference is / why they behave differently even though they both function to request a permission if not already granted.

File: app/src/main/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionRequest.kt [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+        override suspend fun onResult(result: Result.Single) {
+            if (result.granted) {
+                onGranted()
+            }
+        }
+    }
+
+    /**
+     * A notification permission request that shows a bottom sheet prompt before the system dialog.
```
Comment (jpelgrom) [inline app/src/main/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionRequest.kt:123]: This does not show a bottom sheet

File: app/src/main/kotlin/io/homeassistant/companion/android/frontend/session/ServerSessionManager.kt [hunk 8]
```diff
@@ -101,4 +101,46 @@ class ServerSessionManager @Inject constructor(private val serverManager: Server
             RevokeAuthResult.Failed(callbackScript = "${payload.callback}(false)")
         }
     }
+
+    /**
+     * Build the Bearer token authorization header for the given server.
+     *
+     * Used by the download manager to attach credentials to HTTP download requests
+     * made via [android.app.DownloadManager].
+     *
```
Comment (jpelgrom) [inline app/src/main/kotlin/io/homeassistant/companion/android/frontend/session/ServerSessionManager.kt]: Describe usage, don't prescribe when not enforced or necessary. It's not difficult to Ctrl + click to find out where it's used.

```suggestion
```

File: app/src/main/kotlin/io/homeassistant/companion/android/frontend/session/ServerSessionManager.kt [hunk 9]
```diff
@@ -101,4 +101,46 @@ class ServerSessionManager @Inject constructor(private val serverManager: Server
...
+     * @return The Bearer token string, or empty string if unavailable
+     */
+    suspend fun getAuthorizationHeader(serverId: Int): String {
+        return try {
+            serverManager.authenticationRepository(serverId).buildBearerToken()
+        } catch (e: CancellationException) {
+            throw e
+        } catch (e: IllegalStateException) {
+            Timber.e(e, "Failed to build bearer token")
+            ""
```
Comment (jpelgrom) [inline app/src/main/kotlin/io/homeassistant/companion/android/frontend/session/ServerSessionManager.kt]: Why not return `null` so the download can handle an error?

File: app/src/main/kotlin/io/homeassistant/companion/android/frontend/FrontendViewModel.kt [hunk 10]
```diff
@@ -278,6 +283,49 @@ internal class FrontendViewModel @VisibleForTesting constructor(
...
+            )
+            handleDownloadResult(result)
+        }
+    }
+
+    /**
+     * Clears the current pending permission request from the slot.
+     *
+     * Delegates to [PermissionManager.clearPendingPermissionRequest].
+     */
```
Comment (jpelgrom) [inline app/src/main/kotlin/io/homeassistant/companion/android/frontend/FrontendViewModel.kt]: Don't mention the implementation details

```suggestion
    /** Clears the current pending permission request from the slot. */
```

File: app/src/test/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManagerTest.kt [hunk 11]
```diff
@@ -62,53 +79,79 @@ class PermissionManagerTest {
...
+            "true,true,,false", // FCM + granted -> auto-dismiss, no need to prompt
+            "true,false,,true", // FCM + not granted + fresh install -> prompt
+            "true,false,false,false", // FCM + not granted + user already answered -> skip
+            "false,true,,true", // No FCM + granted + fresh install -> prompt to configure websocket
+            "false,false,,true", // No FCM + not granted + fresh install -> prompt
+            "false,false,false,false", // No FCM + not granted + user already answered -> skip
+            "false,true,false,false", // No FCM + granted + user already answered -> skip
         )
-        fun `Given inputs then returns expected result`(
+        fun `Given inputs then sets pending request only when should ask`(
```
Comment (jpelgrom) [inline app/src/test/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManagerTest.kt]: Avoid when because of given/when/then :)

```suggestion
        fun `Given inputs then sets pending request only if it should be asked`(
```

File: app/src/test/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManagerTest.kt [hunk 12]
```diff
@@ -156,202 +205,481 @@ class PermissionManagerTest {
...
+        fun `Given Q+ device then returns false without checking permission`() = runTest {
+            val manager = createManager(sdkInt = Build.VERSION_CODES.Q)
+            val result = manager.requiresStoragePermissionForDownload {}
+
+            assertFalse(result)
+            assertNull(manager.pendingPermissionRequest.value)
+        }
+
+        @Test
+        fun `Given pre-Q device with storage permission already granted then returns false`() = runTest {
```
Comment (jpelgrom) [inline app/src/test/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManagerTest.kt]:
```suggestion
        fun `Given pre-Q device when storage permission already granted then returns false`() = runTest {
```

File: app/src/test/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManagerTest.kt [hunk 13]
```diff
@@ -156,202 +205,481 @@ class PermissionManagerTest {
...
+
+            val manager = createManager(sdkInt = Build.VERSION_CODES.P)
+            val result = manager.requiresStoragePermissionForDownload {}
+
+            assertFalse(result)
+            assertNull(manager.pendingPermissionRequest.value)
+        }
+
+        @Test
+        fun `Given pre-Q device with storage permission not granted then returns true and emits pending request`() = runTest {
```
Comment (jpelgrom) [inline app/src/test/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManagerTest.kt]:
```suggestion
        fun `Given pre-Q device when storage permission not granted then returns true and emits pending request`() = runTest {
```

File: app/src/main/kotlin/io/homeassistant/companion/android/frontend/download/DownloadResult.kt [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+ */
+sealed interface DownloadResult {
+
+    /**
+     * The download was handled, it might have failed but this has been handled properly by the system or the [DataUriDownloadManager].
+     *
+     * For system [android.app.DownloadManager] downloads this means the request was enqueued;
+     * for data URI downloads the file was saved to the Downloads directory.
+     */
+    data object Handled : DownloadResult
```
Comment (jpelgrom) [inline app/src/main/kotlin/io/homeassistant/companion/android/frontend/download/DownloadResult.kt]: Can I propose renaming this to something like `Forwarded`? Handled is very generic; when dispatched to the Javascript interface it's also handled.

Alternatively: change Handled to include more details on how it was handled (system DownloadManager, DataUriDownloadManager, or Javascript). The action for the UI is the same in all cases: nothing.

**Review by jpelgrom — COMMENTED** [review]
This PR tries to do too much at once in my opinion.
 - Reworking permissions might have been necessary but makes the change much more complicated to review
 - `DataUriDownloadManager` + `WebViewActivity`  changes and the new tests for it could be standalone

File: app/src/main/kotlin/io/homeassistant/companion/android/frontend/handler/FrontendMessageHandler.kt (hunk 1 above)
Comment (jpelgrom) [reply app/src/main/kotlin/io/homeassistant/companion/android/frontend/handler/FrontendMessageHandler.kt:65]: I'm fine with both. It is technically more correct but the other name better described what kind of events you should expect here.

File: app/src/main/kotlin/io/homeassistant/companion/android/frontend/download/FrontendDownloadManager.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (jpelgrom) [reply app/src/main/kotlin/io/homeassistant/companion/android/frontend/download/FrontendDownloadManager.kt]: No, didn't expect it but looking at the frontend behavior on my instance the only cookie is `ingress_session` so I can definitely see it being null.

[the remark this answers is not in the document]
Comment (jpelgrom) [reply app/src/main/kotlin/io/homeassistant/companion/android/frontend/download/FrontendDownloadManager.kt]: See: #6709

File: app/src/main/kotlin/io/homeassistant/companion/android/frontend/download/FrontendDownloadManager.kt (hunk 5 above)
Comment (TimoPtr) [reply app/src/main/kotlin/io/homeassistant/companion/android/frontend/download/FrontendDownloadManager.kt:177]: As discussed on Discord https://discord.com/channels/330944238910963714/1346948551892009101/1493694496318292262 we are going to enforce an Opt-In annotation alongside with a refactor to use WebViewActions.

File: app/src/main/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManager.kt (hunk 6 above)
Comment (TimoPtr) [reply app/src/main/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManager.kt]: Renamed to `checkStoragePermissionForDownload` for consistency with the notification and also with the docs that was mentioning checking.

File: app/src/main/kotlin/io/homeassistant/companion/android/frontend/session/ServerSessionManager.kt (hunk 9 above)
Comment (TimoPtr) [reply app/src/main/kotlin/io/homeassistant/companion/android/frontend/session/ServerSessionManager.kt]: Indeed, I copy pasted from the previous code, updating it 👍🏻

File: app/src/main/kotlin/io/homeassistant/companion/android/frontend/download/DownloadResult.kt (hunk 14 above)
Comment (TimoPtr) [reply app/src/main/kotlin/io/homeassistant/companion/android/frontend/download/DownloadResult.kt]: You can I struggled to find a name I liked. I didn't want to add more details now but we could later if needed. I'm going to rename to `Forwarded`.

Comment (TimoPtr) [reply app/src/main/kotlin/io/homeassistant/companion/android/frontend/download/DownloadResult.kt]: I actually merged Dispatched and Forwared since it was actually the same.

File: app/src/main/kotlin/io/homeassistant/companion/android/frontend/download/FrontendDownloadManager.kt (hunk 5 above)
Comment (jpelgrom) [reply app/src/main/kotlin/io/homeassistant/companion/android/frontend/download/FrontendDownloadManager.kt:177]: #6713

**Comment by jpelgrom** [comment]
> This PR tries to do too much at once in my opinion.
> 
>     * Reworking permissions might have been necessary but makes the change much more complicated to review
> 
>     * `DataUriDownloadManager` + `WebViewActivity`  changes and the new tests for it could be standalone

You didn't really address this comment other than mentioning this change has taken some time to complete. I'm not sure splitting it now would make that much of a difference, as it's already in review and other branches are stacked on top of it, but please try to be mindful of it.

**Comment by jpelgrom** [comment]
Testing this I noticed blob downloads aren't working for automation traces, but they are working for camera snapshots... Also fails in latest commit on main with WebViewActivity so out of scope for this PR.

```
2026-04-15 22:27:31.772   406-406   FrontendDownloadManager io....stant.companion.android.debug  D  WebView requested download of blob:https://example.com/ea241ff1-c60e-49e0-b5ff-8fb56d6c2e90
2026-04-15 22:27:31.773   406-406   FrontendDownloadManager io....stant.companion.android.debug  D  Triggering blob download for blob:https://example.com/ea241ff1-c60e-49e0-b5ff-8fb56d6c2e90
2026-04-15 22:27:31.780   406-406   FrontendSc...iewEffects io....stant.companion.android.debug  D  Evaluating script:                 (async function() {
                                                                                                                        try {
                                                                                                                            const response = await fetch("blob:https:\/\/example.com\/ea241ff1-c60e-49e0-b5ff-8fb56d6c2e90");
                                                                                                                            if (!response.ok) {
                                                                                                                                console.error('Blob download failed: HTTP ' + response.status + ' for "blob:https:\/\/example.com\/ea241ff1-c60e-49e0-b5ff-8fb56d6c2e90"');
                                                                                                                                return;
                                                                                                                            }
                                                                                                                            const blob = await response.blob();
                                                                                                                            const reader = new FileReader();
                                                                                                                            reader.onloadend = function() {
                                                                                                        if (typeof window.externalAppV2 !== 'undefined') {
                                                                                                            window.externalAppV2.postMessage(JSON.stringify({type:'externalBus',payload:{type:'handleBlob',data:reader.result,filename:"ea241ff1-c60e-49e0-b5ff-8fb56d6c2e90.json"}}));
                                                                                                        } else {
                                                                                                            window.externalApp.externalBus(JSON.stringify({type:'handleBlob',data:reader.result,filename:"ea241ff1-c60e-49e0-b5ff-8fb56d6c2e90.json"}));
                                                                                                        }
                                                                                                    };
                                                                                                                            reader.readAsDataURL(blob);
                                                                                                                        } catch (e) {
                                                                                                                            console.error('Blob download failed:', e);
                                                                                                                        }
                                                                                                                    })();
2026-04-15 22:27:31.798   406-406   chromium                io....stant.companion.android.debug  I  [INFO:CONSOLE:19] "Blob download failed: TypeError: Failed to fetch", source:  (19)
```

File: app/src/test/kotlin/io/homeassistant/companion/android/frontend/download/FrontendDownloadManagerTest.kt [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+            assertTrue(scriptSlot.captured.contains("type:'handleBlob',data:reader.result"))
+            assertTrue(scriptSlot.captured.contains("fetch("))
+        }
+    }
+
+    @Nested
+    inner class HttpDownload {
+
+        @Test
+        fun `Given http URL with safe credentials when downloadFile called then enqueues download with auth and returns Handled`() = runTest {
```
Comment (jpelgrom) [inline app/src/test/kotlin/io/homeassistant/companion/android/frontend/download/FrontendDownloadManagerTest.kt]:
```suggestion
        fun `Given https URL with safe credentials when downloadFile called then enqueues download with auth and returns Forwarded`() = runTest {
```

File: app/src/test/kotlin/io/homeassistant/companion/android/frontend/download/FrontendDownloadManagerTest.kt [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+                serverId = 1,
+            )
+
+            assertEquals(DownloadResult.Forwarded, result)
+            coVerify { systemDownloadManager.enqueue(any()) }
+            verify { anyConstructed<DownloadManager.Request>().addRequestHeader("Authorization", "Bearer token123") }
+        }
+
+        @Test
+        fun `Given http URL without safe credentials when downloadFile called then enqueues download without auth and returns Handled`() = runTest {
```
Comment (jpelgrom) [inline app/src/test/kotlin/io/homeassistant/companion/android/frontend/download/FrontendDownloadManagerTest.kt]:
```suggestion
        fun `Given https URL without safe credentials when downloadFile called then enqueues download without auth and returns Forwarded`() = runTest {
```

File: app/src/test/kotlin/io/homeassistant/companion/android/frontend/download/FrontendDownloadManagerTest.kt [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+
+            assertEquals(DownloadResult.Forwarded, result)
+            coVerify { systemDownloadManager.enqueue(any()) }
+            verify(exactly = 0) {
+                anyConstructed<DownloadManager.Request>().addRequestHeader("Authorization", any())
+            }
+        }
+
+        @Test
+        fun `Given null system download manager when downloadFile called with http URL then returns Error`() = runTest {
```
Comment (jpelgrom) [inline app/src/test/kotlin/io/homeassistant/companion/android/frontend/download/FrontendDownloadManagerTest.kt]:
```suggestion
        fun `Given null system download manager when downloadFile called with https URL then returns Error`() = runTest {
```

File: app/src/test/kotlin/io/homeassistant/companion/android/frontend/download/FrontendDownloadManagerTest.kt [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+            assertTrue(result is DownloadResult.Error)
+            assertEquals(commonR.string.downloads_failed, (result as DownloadResult.Error).messageResId)
+        }
+    }
+
+    @Nested
+    inner class DataUriDownload {
+
+        @Test
+        fun `Given data URL when downloadFile called then saves via DataUriDownloadManager and returns Handled`() = runTest {
```
Comment (jpelgrom) [inline app/src/test/kotlin/io/homeassistant/companion/android/frontend/download/FrontendDownloadManagerTest.kt]:
```suggestion
        fun `Given data URL when downloadFile called then saves via DataUriDownloadManager and returns Forwarded`() = runTest {
```

File: app/src/test/kotlin/io/homeassistant/companion/android/frontend/download/FrontendDownloadManagerTest.kt [hunk 19]
```diff
@@ -0,0 +1,219 @@
...
+            assertTrue(result is DownloadResult.OpenWithSystem)
+            assertEquals("ftp://example.com/file.txt", (result as DownloadResult.OpenWithSystem).uri.toString())
+        }
+    }
+
+    @Nested
+    inner class HandleBlob {
+
+        @Test
+        fun `Given blob data when handleBlob called then saves via DataUriDownloadManager and returns Handled`() = runTest {
```
Comment (jpelgrom) [inline app/src/test/kotlin/io/homeassistant/companion/android/frontend/download/FrontendDownloadManagerTest.kt]:
```suggestion
        fun `Given blob data when handleBlob called then saves via DataUriDownloadManager and returns Forwarded`() = runTest {
```

File: app/src/test/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManagerTest.kt [hunk 20]
```diff
@@ -62,53 +79,79 @@ class PermissionManagerTest {
...
+        fun `Given pre-TIRAMISU device then does nothing`() = runTest {
+            val manager = createManager(sdkInt = Build.VERSION_CODES.S)
+            manager.checkNotificationPermission(serverId)
+
+            assertNull(manager.pendingPermissionRequest.value)
+        }
     }
 
     @Nested
     inner class OnNotificationPermissionResult {
```
Comment (jpelgrom) [inline app/src/test/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManagerTest.kt]:
```suggestion
    inner class OnPermissionResultNotification {
```

For consistency with the other permission request result inner classes

File: app/src/test/kotlin/io/homeassistant/companion/android/frontend/FrontendViewModelTest.kt [hunk 21]
```diff
@@ -850,4 +820,182 @@ class FrontendViewModelTest {
...
+
+                val event = awaitItem()
+                assertTrue(event is FrontendEvent.OpenExternalLink)
+                assertEquals(testUri, (event as FrontendEvent.OpenExternalLink).uri)
+                cancelAndIgnoreRemainingEvents()
+            }
+        }
+
+        @Test
+        fun `Given handled result when download requested then no UI event emitted`() = runTest {
```
Comment (jpelgrom) [inline app/src/test/kotlin/io/homeassistant/companion/android/frontend/FrontendViewModelTest.kt]:
```suggestion
        fun `Given forwarded result when download requested then no UI event emitted`() = runTest {
```

File: app/src/test/kotlin/io/homeassistant/companion/android/util/DataUriDownloadManagerTest.kt [hunk 22]
```diff
@@ -0,0 +1,356 @@
...
+    fun `Given null filename when saving then notification title uses unnamed file label`() = runTest {
+        val data = Base64.encodeToString("test".toByteArray(), Base64.DEFAULT)
+        val url = "data:text/plain;base64,$data"
+
+        manager.saveDataUri(url = url, mimetype = "text/plain", filename = null)
+
+        val notification = getLatestNotification(url)
+        assertNotNull("Expected a notification to be posted", notification)
+        val title = shadowOf(notification).contentTitle.toString()
+        assertEquals(app.getString(io.homeassistant.companion.android.common.R.string.downloads_unnamed_file), title)
```
Comment (jpelgrom) [inline app/src/test/kotlin/io/homeassistant/companion/android/util/DataUriDownloadManagerTest.kt]: Why not import R as commonR, like we do in a lot of other places?

**Review by jpelgrom — APPROVED** [review]
Testing the changes and reading the code (+ upcoming PRs) I think this is fine, but considering the size I don't guarantee nothing has been missed. Thankfully it's only for the preview and not yet stable.

I think I understand events vs. actions vs. state but look forward to your documentation and edge cases we'll find :)

File: app/src/test/kotlin/io/homeassistant/companion/android/util/DataUriDownloadManagerTest.kt (hunk 22 above)
Comment (TimoPtr) [reply app/src/test/kotlin/io/homeassistant/companion/android/util/DataUriDownloadManagerTest.kt]: Fixed and well I didn't spot it while self reviewing the test that Claude made.

Comment (jpelgrom) [reply app/src/test/kotlin/io/homeassistant/companion/android/util/DataUriDownloadManagerTest.kt]: It happens in multiple places in this file, I only commented on one example

Comment (TimoPtr) [reply app/src/test/kotlin/io/homeassistant/companion/android/util/DataUriDownloadManagerTest.kt]: Addressed

**Review by jpelgrom — APPROVED** [review]
